zio-http/src/main/scala/zhttp/service/Handler.scala

```
@@ -16,40 +16,92 @@ private[zhttp] final case class Handler[R](
+    implicit val iCtx: ChannelHandlerContext = ctx
+    msg match {
+      case jReq: FullHttpRequest =>
+        jReq.touch("server.Handler-channelRead0")
+        unsafeRun(
+          jReq,
+          app,
+          new Request {
+            override def method: Method = Method.fromHttpMethod(jReq.method())
+
+            override def url: URL = URL.fromString(jReq.uri()).getOrElse(null)
+
+            override def headers: Headers = Headers.make(jReq.headers())
+
+            override def remoteAddress: Option[InetAddress] = {
+              ctx.channel().remoteAddress() match {
+                case m: InetSocketAddress => Some(m.getAddress)
+                case _                    => None
+              }
+            }
+
+            override def data: HttpData = HttpData.fromByteBuf(jReq.content())
+          },
+        )
+      case jReq: HttpRequest     =>
+        unsafeRun(
+          jReq,
+          app,
+          new Request {
+            override def data: HttpData = HttpData.Incoming(callback =>
+              ctx
+                .pipeline()
+                .addAfter(HTTP_REQUEST_HANDLER, HTTP_CONTENT_HANDLER, new RequestBodyHandler(callback)): Unit,
+            )
+
+            override def headers: Headers = Headers.make(jReq.headers())
+
+            override def method: Method = Method.fromHttpMethod(jReq.method())
+
+            override def remoteAddress: Option[InetAddress] = {
+              ctx.channel().remoteAddress() match {
+                case m: InetSocketAddress => Some(m.getAddress)
+                case _                    => None
+              }
+            }
+
+            override def url: URL = URL.fromString(jReq.uri()).getOrElse(null)
+          },
+        )
+
+      case msg: HttpContent => ctx.fireChannelRead(msg): Unit
+
+      case _ =>
+        ctx.writeAndFlush(new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NOT_ACCEPTABLE)): Unit
```

We should throw `InvalidStateException` here.